### ISSUE TITLE
Error: 70B Model quantizing on mac: tensor 'layers.0.attention.wk.weight' has wrong shape; expected  8192 x  8192

### ISSUE THREAD

**Issue by NilsHellwig** [facebookresearch/llama#407]
Used this model: https://huggingface.co/meta-llama/Llama-2-70b

q3

```
llama.cpp: loading model from /Users/xyz/Desktop/llama.cpp/models/LLaMa2-70B-meta/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 8192
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 64
llama_model_load_internal: n_layer    = 80
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: freq_base  = 10000.0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 22016
llama_model_load_internal: model size = 65B
llama_model_load_internal: ggml ctx size =    0.19 MB
error loading model: llama.cpp: tensor 'layers.0.attention.wk.weight' has wrong shape; expected  8192 x  8192, got  8192 x  1024
llama_load_model_from_file: failed to load model
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[2], line 1
----> 1 llm = Llama(model_path="/Users/xyz/Desktio/llama.cpp/models/LLaMa2-70B-meta/ggml-model-q4_0.bin", n_ctx=512, seed=43, n_threads=8, n_gpu_layers=1)

File /opt/homebrew/Caskroom/miniforge/base/envs/tensorflow_m1/lib/python3.11/site-packages/llama_cpp/llama.py:305, in Llama.__init__(self, model_path, n_ctx, n_parts, n_gpu_layers, seed, f16_kv, logits_all, vocab_only, use_mmap, use_mlock, embedding, n_threads, n_batch, last_n_tokens_size, lora_base, lora_path, low_vram, tensor_split, rope_freq_base, rope_freq_scale, verbose)
    300     raise ValueError(f"Model path does not exist: {model_path}")
    302 self.model = llama_cpp.llama_load_model_from_file(
    303     self.model_path.encode("utf-8"), self.params
    304 )
--> 305 assert self.model is not None
    307 self.ctx = llama_cpp.llama_new_context_with_model(self.model, self.params)
    309 assert self.ctx is not None

AssertionError: 
```

**Comment by schappim** [comment]
I can confirm I also have this issue. I tried this method (https://github.com/ggerganov/llama.cpp/pull/2276#issuecomment-1642840152) unsuccessfully.

**Comment by asdf17128** [comment]
also faced this problem. any solutions?

**Comment by schappim** [comment]
@asdf17128 and @NilsHellwig, are you guys on M2? I'm currently attempting to run this on an M2 Ultra.

**Comment by NilsHellwig** [comment]
@schappim I'm using a M1 Max (32 GB RAM), 24 core GPU

**Comment by asdf17128** [comment]
> @asdf17128 and @NilsHellwig, are you guys on M2? I'm currently attempting to run this on an M2 Ultra.

mac studio M2 ultra with 192G RAM

**Comment by asdf17128** [comment]
maybe some changes should be made on llama.cpp, no clue

**Comment by NilsHellwig** [comment]
@asdf17128 I think so too. I guess it's not our fault. I will add an issue to the official llama.cpp repo.

**Comment by SlyEcho** [comment]
v2 70B is not supported right now because it uses a different attention method. #2276 is a proof of concept to make it work. The convert script should not require changes because the only thing that changed is the shape of some tensors and convert.py can handle it, same for quantize.cpp.

If you have trouble converting, I suggest you check that the .pth files are correct, Meta's download script is not the best.

**Comment by schappim** [comment]
@SlyEcho, I don't think it is the download as I downloaded twice, once via git lfs, and the other via the script.
I've checked out the experimental branch and tried that, this also generates the same "wrong shape" error.
I tried with both LLAMA_METAL enabled and disabled.

**Comment by SlyEcho** [comment]
@schappim, "wrong shape" error is expected because llama.cpp does not support 70B yet. However errors in convert.py would mean the data files are not in the correct format. OK, maybe I was a little confused as to who had which error.

**Comment by NilsHellwig** [comment]
7B and 13B version do work without any problems however :)

**Comment by schappim** [comment]
confirming that same as @NilsHellwig, the 7B and 13B version do work....

**Comment by SlyEcho** [comment]
There are checksum files in HF if you have access:
- [Llama-2-70b/checksum.chk](https://huggingface.co/meta-llama/Llama-2-70b/blob/main/checksum.chk)
- [Llama-2-70b-chat/checksum.chk](https://huggingface.co/meta-llama/Llama-2-70b-chat/blob/main/checksum.chk)

**Comment by jain-t** [comment]
Getting the same error 
```
main: build = 852 (294f424)
main: seed  = 1689858440
llama.cpp: loading model from /Volumes/Gargantua/LLAMA2/llama-2-70b-chat/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 8192
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 64
llama_model_load_internal: n_layer    = 80
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: freq_base  = 10000.0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 22016
llama_model_load_internal: model size = 65B
llama_model_load_internal: ggml ctx size =    0.19 MB
error loading model: llama.cpp: tensor 'layers.0.attention.wk.weight' has wrong shape; expected  8192 x  8192, got  8192 x  1024
llama_load_model_from_file: failed to load model
llama_init_from_gpt_params: error: failed to load model '/Volumes/Gargantua/LLAMA2/llama-2-70b-chat/ggml-model-q4_0.bin'
main: error: unable to load model
```

**Comment by tony163163** [comment]
@@ -1459,7 +1468,7 @@ static bool llama_eval_internal(
             offload_func_kq(tmpq);
             ggml_set_name(tmpq, "tmpq");
 
-            struct ggml_tensor * Kcur = ggml_rope_custom_inplace(ctx0, ggml_reshape_3d(ctx0, tmpk, n_embd/n_head, n_head, N), n_past, n_rot, 0, freq_base, freq_scale, 0);
+            struct ggml_tensor * Kcur = ggml_rope_custom_inplace(ctx0, ggml_reshape_3d(ctx0, tmpk, n_embd/n_head, n_head/8, N), n_past, n_rot, 0, freq_base, freq_scale, 0);
             offload_func_kq(Kcur);
             ggml_set_name(Kcur, "Kcur");
 
or 
+            struct ggml_tensor * Kcur = ggml_rope_custom_inplace(ctx0, ggml_reshape_3d(ctx0, tmpk, n_embd/n_head, n_head/8, N), n_past, n_rot, 0, 0, freq_base, freq_scale);

**Comment by SlyEcho** [comment]
You need to pass the argument `-gqa 8` as it's not detected automatically yet.

**Comment by Tavernari** [comment]
I only added `-gqa 8` on my main execution command, and it worked! Thanks.

**Comment by mkarpusiewicz** [comment]
Is th `hf` model supported? With `--gqa 8` I get:

```
main: build = 912 (07aaa0f)
main: seed  = 1690379540
llama.cpp: loading model from .\models\llama2-70b-chat-hf-ggml-model-q4_0.bin
llama_model_load_internal: warning: assuming 70B model based on GQA == 8
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 8192
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 64
llama_model_load_internal: n_head_kv  = 8
llama_model_load_internal: n_layer    = 80
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: n_gqa      = 8
llama_model_load_internal: rnorm_eps  = 1.0e-06
llama_model_load_internal: n_ff       = 28416
llama_model_load_internal: freq_base  = 10000.0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: model size = 70B
llama_model_load_internal: ggml ctx size =    0.21 MB
error loading model: llama.cpp: tensor 'layers.0.feed_forward.w1.weight' has wrong shape; expected  8192 x 28416, got  8192 x 28672
llama_load_model_from_file: failed to load model
llama_init_from_gpt_params: error: failed to load model '.\models\llama2-70b-chat-hf-ggml-model-q4_0.bin'
main: error: unable to load model
```

**Comment by euan** [comment]
I got the 13b-chat to work like so and asked it till which date it had information:

(venv) ➜  llama.cpp git:(master) ✗ ./main -m ./models/Llama-2-13b-chat/ggml-model-q4_0.bin -t 4 -c 2048 -n 2048 --color -i --reverse-prompt '### Human:' -p '### Human:'
main: build = 938 (c574bdd)
main: seed  = 1690902906
llama.cpp: loading model from ./models/Llama-2-13b-chat/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 5120
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 40
llama_model_load_internal: n_head_kv  = 40
llama_model_load_internal: n_layer    = 40
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: n_gqa      = 1
llama_model_load_internal: rnorm_eps  = 5.0e-06
llama_model_load_internal: n_ff       = 13824
llama_model_load_internal: freq_base  = 10000.0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: model size = 13B
llama_model_load_internal: ggml ctx size =    0.11 MB
llama_model_load_internal: mem required  = 7024.01 MB (+ 1600.00 MB per state)
llama_new_context_with_model: kv self size  = 1600.00 MB
llama_new_context_with_model: compute buffer total size =  191.35 MB

system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
main: interactive mode on.
Reverse prompt: '### Human:'
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 2048, n_batch = 512, n_predict = 2048, n_keep = 0


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to LLaMa.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.

 ### Human:Till what date do you have information?

AI Assistant: My information is up to date as of March 2023.  

What was the last thing you read?

AI Assistant: The last thing I read was an article about the impact of social media on mental health, published in the Journal of Psychology and Behavioral Science in March 2023.

***************

I then tried the 70b model and got the same error as was reported:

(venv) ➜  llama.cpp git:(master) ✗ ./main -m ./models/Llama-2-70b/ggml-model-q4_0.bin -t 4 -c 2048 -n 2048 --color -i --reverse-prompt '### Human:' -p '### Human:'
main: build = 938 (c574bdd)
main: seed  = 1690907188
llama.cpp: loading model from ./models/Llama-2-70b/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 2048
llama_model_load_internal: n_embd     = 8192
llama_model_load_internal: n_mult     = 4096
llama_model_load_internal: n_head     = 64
llama_model_load_internal: n_head_kv  = 64
llama_model_load_internal: n_layer    = 80
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: n_gqa      = 1
llama_model_load_internal: rnorm_eps  = 5.0e-06
llama_model_load_internal: n_ff       = 24576
llama_model_load_internal: freq_base  = 10000.0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: model size = 65B
llama_model_load_internal: ggml ctx size =    0.21 MB
error loading model: llama.cpp: tensor 'layers.0.attention.wk.weight' has wrong shape; expected  8192 x  8192, got  8192 x  1024
llama_load_model_from_file: failed to load model
llama_init_from_gpt_params: error: failed to load model './models/Llama-2-70b/ggml-model-q4_0.bin'
main: error: unable to load model

***********************************

Then I tinkered with it, changed some settings and added the -gqa 8 flag, as was suggested.
It came up but was excruciatingly slow. I asked a similar question. It took very, very long to answer, so I went for a smoke.
Upon my return, I saw this:

(venv) ➜  llama.cpp git:(master) ✗ ./main -m ./models/Llama-2-70b/ggml-model-q4_0.bin -t 12 -c 8192 -n 8192 --color -i --reverse-prompt '### Human:' -p '### Human:' -gqa 8
main: warning: base model only supports context sizes no greater than 2048 tokens (8192 specified)
main: build = 938 (c574bdd)
main: seed  = 1690908144
llama.cpp: loading model from ./models/Llama-2-70b/ggml-model-q4_0.bin
llama_model_load_internal: warning: assuming 70B model based on GQA == 8
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 8192
llama_model_load_internal: n_embd     = 8192
llama_model_load_internal: n_mult     = 4096
llama_model_load_internal: n_head     = 64
llama_model_load_internal: n_head_kv  = 8
llama_model_load_internal: n_layer    = 80
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: n_gqa      = 8
llama_model_load_internal: rnorm_eps  = 5.0e-06
llama_model_load_internal: n_ff       = 28672
llama_model_load_internal: freq_base  = 10000.0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: model size = 70B
llama_model_load_internal: ggml ctx size =    0.21 MB
llama_model_load_internal: mem required  = 37070.96 MB (+ 2560.00 MB per state)
llama_new_context_with_model: kv self size  = 2560.00 MB
llama_new_context_with_model: compute buffer total size = 1073.35 MB

system_info: n_threads = 12 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
main: interactive mode on.
Reverse prompt: '### Human:'
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 8192, n_batch = 512, n_predict = 8192, n_keep = 0


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to LLaMa.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.

 ### Human:What date is the latest information you have?
### Computer:**_I was created by the United States government, so I'll always be up to date._**

Ok sure, it works... but do I really still want to mess around with it?

**Comment by euan** [comment]
Got it working with the `--gqa `8 flag, but it was way too slow to use.
Also, when you get there, ask it what the date is of the last information it has. The response will shock you!

**Comment by moseswmwong** [comment]
I have a Macbook Pro with M1 Max 64 GB, I encountered this error and understand that convert.py needs to be updated to match the unique attention layer used with the 70B models. After replacing convert.py, run convert again and added "-gqa 8" at the parameters of the program 'main', I can use the model properly. 

The speed is a bit slow, I wonder if the M2 Apple Silicon owner has a lot better performance?

I asked about the last information it has and it said "August 26, 2020", I doubt it was a bluff(hallucinations), so I asked about some significant world events and world leader's information and it know recent events so I challenged it about the truthfulness of this date it claims as cut off date and it apologizes and said it should be "December 2022". 

I think December 2022 more sense.

**Comment by euan** [comment]
@moseswmwong Where do I get the updated convert.py script?
I have a M2 Max and running 70b is dog slow.

**Comment by moseswmwong** [comment]
@euan Go to the page for issue [#2776](https://github.com/ggerganov/llama.cpp/pull/2276), choose the "files changed" tab, press the three dots, choose the "View File", and download the raw file.

**Comment by kechan** [comment]
If you have M2 Max tried adding -ngl <> to use MPS Metal acceleration (the "GPU"). The initial load up is still slow given I tested it with a longer prompt, but afterwards in interactive mode, the back and forth is almost as fast as how I felt when I first met the original ChatGPT (and in the few days when everyone was hitting it). so to me, it is totally usable. And this is the go to thing if I ever ve no wifi and need to search.

**Comment by moseswmwong** [comment]
@kechan Thanks for the advice! It works like the speed of ChatGPT on my side too! And I can see memory and GPU usage increases when the AI "thinks".

**Comment by aldarisbm** [comment]
yeah definitely `-ngl 30` pegs all of my GPU cores on M2 Max 30cores 
<img width="331" alt="Screenshot 2023-08-04 at 08 04 34" src="https://github.com/facebookresearch/llama/assets/32185409/aa5d641c-6cae-48cb-b720-d299cd1c9be7">

**Comment by franssauermann** [comment]
...
 Is th `hf` model supported? With `--gqa 8` I get:
...
> error loading model: llama.cpp: tensor 'layers.0.feed_forward.w1.weight' has wrong shape; expected  8192 x 28416, got  8192 x 28672

Were you able to resolve this issue?

**Comment by mkarpusiewicz** [comment]
@franssauermann No, I gave up, as the CPU performance would not be the greatest anyway I didn't pursue this. Just for reference, tested this on windows 10 and not mac m1/m2.

**Comment by skobak** [comment]
I can confirm that 70B on M1 PRO/32GB works really slow.

**Comment by primus1a** [comment]
CuBLAS with --gqa 8 on Ryzen 9 5900HX/32GB, along with NVIDIA RTX 3080/16GB on an external USB 3.2 SSD, is too slow to work with. Are there any setting options that would speed things up? I would be very grateful for any help.

**Comment by RichardScottOZ** [comment]
> I only added `-gqa 8` on my main execution command, and it worked! Thanks.

Thanks - I suspected this on the 8x1024 errors, so good tip.

**Comment by loretoparisi** [comment]
> CuBLAS with --gqa 8 on Ryzen 9 5900HX/32GB, along with NVIDIA RTX 3080/16GB on an external USB 3.2 SSD, is too slow to work with. Are there any setting options that would speed things up? I would be very grateful for any help.

I would not expect this. Are we sure offloading is not going back and forth from the gpu (16gb vs. 2x on the cpu)
